src/App.tsx

```diff
@@ -1,24 +1,239 @@
+    getTodos(USER_ID)
+      .then(setAllTodos)
+      .catch(() => pushError(ErrorTypes.Load));
+  }, []);
+
+  const filterHandler = (array: Todo[], filterType: string) => {
+    switch (filterType) {
+      case Filter.Active:
+        return array.filter(item => !item.completed);
+      case Filter.Completed:
+        return array.filter(item => item.completed);
+      default:
+        return array;
+    }
+  };
+
+  const addTodoHandler = useCallback(() => {
+    if (!title.trim()) {
+      pushError(ErrorTypes.Add);
+
+      return;
+    }
+
+    setIsAdding(true);
+
+    const todo = {
+      title,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo({
+      id: 0,
+      title,
+      userId: USER_ID,
+      completed: false,
+    });
+
+    addTodo(todo)
+      .then(result => {
+        setAllTodos(
+          prevTodos => [...prevTodos, result],
+        );
+        setTempTodo(null);
+      })
+      .catch(() => {
+        setError(ErrorTypes.Add);
+        setTempTodo(null);
+      })
+      .finally(() => {
+        setIsAdding(false);
+        setTitle('');
+        setTempTodo(null);
+      });
+  }, [title]);
+
+  const deleteTodoHandler = useCallback((id: number) => {
+    deleteTodo(id)
+      .then(() => setAllTodos(
+        prevAllTodos => prevAllTodos.filter(prevTodo => prevTodo.id !== id),
+      ))
+      .catch(() => pushError(ErrorTypes.Delete));
+  }, []);
+
+  const activeTodos = useMemo(
+    () => filterHandler(allTodos, Filter.Active), [allTodos],
+  );
+  const completedTodos = useMemo(
+    () => filterHandler(allTodos, Filter.Completed), [allTodos],
+  );
+  const visibleTodos = useMemo(
+    () => filterHandler(allTodos, filter), [allTodos, filter],
+  );
+
+  const deleteAllCompletedHandler = useCallback(() => {
+    completedTodos.map(
+      completedTodo => deleteTodoHandler(completedTodo.id),
+    );
+  }, [allTodos]);
+
+  const todoStatusChangeHandler = useCallback(
+    (id: number, isCompeled: boolean) => {
+      setSelectedTodoIds(prev => [...prev, id]);
+
+      completeTodo(id, isCompeled)
+        .then(() => setAllTodos(prevTodos => {
+          const todosCopy = JSON.parse(JSON.stringify(prevTodos));
+
+          return todosCopy.map(
+            (todo: Todo) => (
+              todo.id === id ? { ...todo, completed: isCompeled } : todo
+            ),
+          );
+        }))
+        .catch(() => pushError(ErrorTypes.Update))
+        .finally(() => setSelectedTodoIds([]));
+    }, [],
+  );
+
+  const completeAll = useCallback(() => {
+    Promise.all(allTodos.map(
+      todo => todoStatusChangeHandler(
+        todo.id, !allTodos.every(todoItem => todoItem.completed),
+      ),
+    ));
+  }, [allTodos]);
+
+  const doubleClickHandler = useCallback((id: number) => {
+    setSelectedTodoIds([id]);
+    setIsEditing(true);
+  }, []);
+
+  const cancelEditingHandler = useCallback(() => {
+    setIsEditing(false);
+    setSelectedTodoIds([]);
+  }, []);
```

Comment:
You don't need to change it now. But read it for the future, please 
Memoization is a costly thing for memory. Using it for every component rendering might decrease performance since it stores results that increase the memory used for the program. We should only use memoization when there is a clear benefit.
You don't really need to use it for simple values or functions. If you’re interested in it, you can read more about it [here](https://kentcdodds.com/blog/usememo-and-usecallback), for example